rslib/src/decks/addupdate.rs

```diff
@@ -43,9 +43,7 @@ impl Collection {
     }
 
     pub(crate) fn add_deck_inner(&mut self, deck: &mut Deck, usn: Usn) -> Result<()> {
-        if deck.id.0 != 0 {
-            return Err(AnkiError::invalid_input("deck to add must have id 0"));
-        }
+        ensure_valid_input!(deck.id.0 == 0, "deck to add must have id 0");
```

Comment:
I know what you mean and had also considered the inverse variant. But my best guess is that after getting familiar with this, it will actually be more intuitive. After all, it basically works like any assert statement. And if you work with the code below, I think you look up to see which contraints hold true more frequently than to see which cases are excluded.

Reply:
Fair point about assert, obviously there is some bias here from what I'm used to. I think I could adjust to mentally map ensure->assert fairly easily - "(ensure/assert) (condition) (or failure message)". I find the code above a little harder to parse, since "valid_input" is placed between the ensure/assert and the condition. I'd probably get used to it, but can't help but wonder if there could be a better wording.

Throwing out some other ideas:

```
invalid_if_not!(deck.id.0 == 0, "deck to add must have id 0");
```

(but then inverting the conditional would probably make things simpler)

Or

```
ensuring!(deck.id.0 == 0, "deck to add must have id 0");
```

Or maybe even implementing the invalid trait for bool:

```
(deck_id.0 == 0).or_invalid("deck to add must have id 0")?;
```

Thoughts?

Reply:
Implementing on bool is a cool idea, but it would prevent the usage of format strings.
I don't mind `ensuring!`, but how about `validate!`? Maybe that would link better to the name of the error.

Reply:
I don't mind validate, but feel it doesn't convey "must be true" quite as well as other words like assert/ensure. What about other similar words like require/enforce/check/assume?

Reply:
I'll go with `require` then. 🙂

Reply:
There's precedent :-) 

https://en.wikipedia.org/wiki/Precondition

(I actually took a few courses back in my university days that used Eiffel)